SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -0,0 +1,398 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Management;
+using System.Net.NetworkInformation;
+using System.Text;
+using Microsoft.Win32;
+using SteamKit2.Util.MacHelpers;
+
+using static SteamKit2.Util.MacHelpers.LibC;
+using static SteamKit2.Util.MacHelpers.CoreFoundation;
+using static SteamKit2.Util.MacHelpers.DiskArbitration;
+using static SteamKit2.Util.MacHelpers.IOKit;
+using System.Threading.Tasks;
+
+namespace SteamKit2
+{
+    abstract class MachineInfoProvider
+    {
+        public static MachineInfoProvider GetProvider()
+        {
+            switch ( Environment.OSVersion.Platform )
+            {
+                case PlatformID.Win32NT:
+                case PlatformID.Win32Windows:
+                    return new WindowsInfoProvider();
+
+                case PlatformID.Unix:
+                    if ( Utils.IsRunningOnDarwin() )
+                    {
+                        return new OSXInfoProvider();
+                    }
+                    else
+                    {
+                        return new LinuxInfoProvider();
+                    }
+            }
+
+            return new DefaultInfoProvider();
+        }
+
+        public abstract byte[] GetMachineGuid();
+        public abstract byte[] GetMacAddress();
+        public abstract byte[] GetDiskId();
+    }
+
+    class DefaultInfoProvider : MachineInfoProvider
+    {
+        public override byte[] GetMachineGuid()
+        {
+            return Encoding.UTF8.GetBytes( Environment.MachineName + "-SteamKit" );
+        }
+
+        public override byte[] GetMacAddress()
+        {
+            // mono seems to have a pretty solid implementation of NetworkInterface for our platforms
+            // if it turns out to be buggy we can always roll our own and poke into /sys/class/net on nix
+
+            var firstEth = NetworkInterface.GetAllNetworkInterfaces()
+                .Where( i => i.NetworkInterfaceType == NetworkInterfaceType.Ethernet )
```

Comment:
>and `Wireless80211` for WiFi.

Interesting, the [MSDN docs](https://msdn.microsoft.com/en-us/library/system.net.networkinformation.networkinterface.networkinterfacetype%28v=vs.100%29.aspx) don't state that that's a possible value. I suppose it can't hurt to include wireless adapters as well.

On Windows, Steam is using [`GetAdaptersInfo`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365917%28v=vs.85%29.aspx) and checks for adapters of type `MIB_IF_TYPE_ETHERNET` and `MIB_IF_TYPE_OTHER`.